File: spec/models/clever_questions/select_documents_form_spec.rb
require 'spec_helper'
require 'rails_helper'

describe CleverQuestions::SelectDocumentsForm do
  context '#validations' do
    context '#invalid form' do
      it 'should be invalid if all inputs are empty' do
        form = CleverQuestions::SelectDocumentsForm.new({})
        expect(form).to_not be_valid
        expect(form.errors.full_messages).to eql ['Please select the documents you have']
      end

      it 'should be invalid if no driving licence details are given' do
        form = CleverQuestions::SelectDocumentsForm.new(
          any_driving_licence: 'true',
          passport: 'false'
        )
        expect(form).to_not be_valid
        expect(form.errors.full_messages).to eql ['Please select the documents you have']
      end

      it 'should be invalid if user only inputs driving licence details' do
        form = CleverQuestions::SelectDocumentsForm.new(
          any_driving_licence: 'true',
          driving_licence: 'true'
        )
        expect(form).to_not be_valid
        expect(form.errors.full_messages).to eql ['Please select the documents you have']
      end

      it 'should be invalid if user only inputs passport details' do
        form = CleverQuestions::SelectDocumentsForm.new(
          passport: 'true'
        )
        expect(form).to_not be_valid
        expect(form.errors.full_messages).to eql ['Please select the documents you have']
      end

      context 'passport expiry' do
        it 'should be invalid if passport expiry date not present when passport has expired' do
          form = CleverQuestions::SelectDocumentsForm.new(
            any_driving_licence: 'false',
            ni_driving_licence: 'true',
            driving_licence: 'true',
            passport: 'yes_expired'
          )

          expect(form).to_not be_valid
          expect(form.errors.full_messages).to eql ['Please enter valid passport expiry date']
        end

        it 'should be invalid if passport expiry date day element not present' do
          form = select_documents_form_with_passport_expiry(
            day: '',
            month: '1',
            year: '2017'
          )

          expect(form).to_not be_valid
          expect(form.errors.full_messages).to eql ['Please enter valid passport expiry date']
        end

        it 'should be invalid if passport expiry date month element not present' do
          form = select_documents_form_with_passport_expiry(
            day: '1',
            month: '',
            year: '2017'
          )

          expect(form).to_not be_valid
          expect(form.errors.full_messages).to eql ['Please enter valid passport expiry date']
        end

        it 'should be invalid if passport expiry date year element not present' do
          form = select_documents_form_with_passport_expiry(
            day: '1',
            month: '1',
            year: ''
           )

          expect(form).to_not be_valid
          expect(form.errors.full_messages).to eql ['Please enter valid passport expiry date']
        end

        it 'should be invalid if passport expiry date day element is less than one' do
          form = select_documents_form_with_passport_expiry(
            day: '0',
            month: '1',
            year: '2017'
          )

          expect(form).to_not be_valid
          expect(form.errors.full_messages).to eql ['Please enter valid passport expiry date']
        end

        it 'should be invalid if passport expiry date day element is greater than 31' do
          form = select_documents_form_with_passport_expiry(
            day: '32',
            month: '1',
            year: '2016'
          )

          expect(form).to_not be_valid
          expect(form.errors.full_messages).to eql ['Please enter valid passport expiry date']
        end

        it 'should be invalid if passport expiry date day element is less than than 1' do
          form = select_documents_form_with_passport_expiry(
            day: '-1',
            month: '1',
            year: '2016'
          )

          expect(form).to_not be_valid
          expect(form.errors.full_messages).to eql ['Please enter valid passport expiry date']
        end

        it 'should be invalid if passport expiry date day element is not an integer' do
          form = select_documents_form_with_passport_expiry(
            day: 'aaa',
            month: '1',
            year: '2016'
          )

          expect(form).to_not be_valid
          expect(form.errors.full_messages).to eql ['Please enter valid passport expiry date']
        end

        it 'should be invalid if passport expiry date month element is less than one' do
          form = select_documents_form_with_passport_expiry(
            day: '1',
            month: '-1',
            year: '2016'
          )

          expect(form).to_not be_valid
          expect(form.errors.full_messages).to eql ['Please enter valid passport expiry date']
        end

        it 'should be invalid if passport expiry date month element is more than twelve' do
          form = select_documents_form_with_passport_expiry(
            day: '1',
            month: '14',
            year: '2016'
           )

          expect(form).to_not be_valid
          expect(form.errors.full_messages).to eql ['Please enter valid passport expiry date']
        end

        it 'should be invalid if passport expiry date month element is not an integer' do
          form = select_documents_form_with_passport_expiry(
            day: '1',
            month: 'aaa',
            year: '2016'
          )

          expect(form).to_not be_valid
          expect(form.errors.full_messages).to eql ['Please enter valid passport expiry date']
        end

        it 'should be invalid if passport expiry date year element is more than this year' do
          form = select_documents_form_with_passport_expiry(
            day: '1',
            month: '1',
            year: (Date.today.year + 1).to_s
        )

          expect(form).to_not be_valid
          expect(form.errors.full_messages).to eql ['Please enter valid passport expiry date']
        end

        it 'should be invalid if passport expiry date year element is less than zero' do
          form = select_documents_form_with_passport_expiry(
            day: '1',
            month: '1',
            year: '-1'
          )

          expect(form).to_not be_valid
          expect(form.errors.full_messages).to eql ['Please enter valid passport expiry date']
        end

        it 'should be invalid if passport expiry date year element is not an integer' do
          form = select_documents_form_with_passport_expiry(
            day: '1',
            month: '1',
            year: 'aaa'
          )

          expect(form).to_not be_valid
          expect(form.errors.full_messages).to eql ['Please enter valid passport expiry date']
        end
      end
    end

    context '#valid form' do
      it 'should be valid if answers are given to every question if passpport not expired' do
        form = CleverQuestions::SelectDocumentsForm.new(
          any_driving_licence: 'false',
          ni_driving_licence: 'true',
          driving_licence: 'true',
          passport: 'true',
        )
        expect(form).to be_valid
      end

      it 'should set passport expiry details to empty if not supplied' do
        form = CleverQuestions::SelectDocumentsForm.new(
          any_driving_licence: 'false',
          ni_driving_licence: 'true',
          driving_licence: 'true',
          passport: 'true',
        )
        expect(form.passport_expiry).to eql(day: '', month: '', year: '')
      end

      it 'should be valid if answers are given to every question if passport expired' do
        form = CleverQuestions::SelectDocumentsForm.new(
          any_driving_licence: 'false',
          ni_driving_licence: 'true',
          driving_licence: 'true',
          passport: 'yes_expired',
          passport_expiry: {
            day: Date.today.day.to_s,
            month: Date.today.month.to_s,
            year: Date.today.year.to_s
          }
        )
        expect(form).to be_valid
      end
    end
  end

  context '#selected_answers' do
    it 'should return a hash of the selected answers' do
      form = CleverQuestions::SelectDocumentsForm.new(
        passport: 'false',
      )
      evidence = form.selected_answers
      expect(evidence).to eql(passport: false)
    end

    it 'should return a hash of the no driving licence and no ni driving licence if no selected for any driving licence' do
      form = CleverQuestions::SelectDocumentsForm.new(
        any_driving_licence: 'false',
      )
      evidence = form.selected_answers
      expect(evidence).to eql(driving_licence: false, ni_driving_licence: false)
    end

    it 'should return a hash of driving licence true if GB driving licence selected' do
      form = CleverQuestions::SelectDocumentsForm.new(
        any_driving_licence: 'true',
        driving_licence: 'true'
      )
      evidence = form.selected_answers
      expect(evidence).to eql(driving_licence: true)
    end

    it 'should not return selected answers when there it is not an eligible IDP evidence ' do
      form = CleverQuestions::SelectDocumentsForm.new(
        passport: 'true',
        any_driving_licence: ''
      )
      answers = form.selected_answers
      expect(answers).to eql(passport: true)
    end

    it 'has passport if expired under 6 months' do
      date_in_past = Date.today - 2.months
      form = CleverQuestions::SelectDocumentsForm.new(
        passport: 'yes_expired',
        passport_expiry: {
          day: date_in_past.day.to_s,
          month: date_in_past.month.to_s,
          year: date_in_past.year.to_s
        }
      )
      evidence = form.selected_answers
      expect(evidence).to eql(passport: true)
    end

    it 'has passport if expired exactly 6 months' do
      date_in_past = Date.today - 6.months
      form = CleverQuestions::SelectDocumentsForm.new(
        passport: 'yes_expired',
        passport_expiry: {
          day: date_in_past.day.to_s,
          month: date_in_past.month.to_s,
          year: date_in_past.year.to_s
        }
      )
      evidence = form.selected_answers
      expect(evidence).to eql(passport: true)
    end

    it 'has no passport if expired over 6 months' do
      date_in_past = Date.today - 7.months
      form = CleverQuestions::SelectDocumentsForm.new(
        passport: 'yes_expired',
        passport_expiry: {
          day: date_in_past.day.to_s,
          month: date_in_past.month.to_s,
          year: date_in_past.year.to_s
        }
      )
      evidence = form.selected_answers
      expect(evidence).to eql(passport: false)
    end

    it 'has no passport if expired 6 months and 1 day' do
      date_in_past = Date.today - 6.months - 1.day
      form = CleverQuestions::SelectDocumentsForm.new(
        passport: 'yes_expired',
        passport_expiry: {
          day: date_in_past.day.to_s,
          month: date_in_past.month.to_s,
          year: date_in_past.year.to_s
        }
      )
      evidence = form.selected_answers
      expect(evidence).to eql(passport: false)
    end
  end

  context '#further identity information' do
    it 'should require further information when user has neither uk passport or driving licence' do
      form = CleverQuestions::SelectDocumentsForm.new(
        any_driving_licence: 'false',
        passport: 'false'
      )
      expect(form).to be_further_id_information_required
    end

    it 'should not require further information when user has a northern ireland driving licence' do
      form = CleverQuestions::SelectDocumentsForm.new(
        any_driving_licence: 'true',
        ni_driving_licence: 'true',
        passport: 'false'
      )
      expect(form).to_not be_further_id_information_required
    end

    it 'should not require further information when user has a GB driving licence' do
      form = CleverQuestions::SelectDocumentsForm.new(
        any_driving_licence: 'true',
        driving_licence: 'true',
        passport: 'false'
      )
      expect(form).to_not be_further_id_information_required
    end

    it 'should not require further information when user has a UK passport' do
      form = CleverQuestions::SelectDocumentsForm.new(
        any_driving_licence: 'false',
        passport: 'true'
      )
      expect(form).to_not be_further_id_information_required
    end
  end

private

  def select_documents_form_with_passport_expiry(passport_expiry)
    CleverQuestions::SelectDocumentsForm.new(
      any_driving_licence: 'false',
      ni_driving_licence: 'true',
      driving_licence: 'true',
      passport: 'yes_expired',
      passport_expiry: passport_expiry
    )
  end
end

File: spec/javascripts/clever_questions/select_documents_form_spec.js
describe("Select Documents Form", function () {

    var formWithNoErrors = '<div id="no-documents-message" class="visually-hidden" aria-live="assertive"></div>' +
        '<div class="grid-row">' +
        '<div class="column-two-thirds">' +
        '<h1 class="heading-large">Your photo identity document</h1>' +
        '<p>Certified companies use information from identity documents to verify you.</p>' +
        '<div class="panel panel-border-narrow">' +
        'The more identity documents you can provide now, the more likely it is that the company can verify you successfully.' +
        '</div>' +
        '<form id="validate-select-documents" class="select-documents-form" novalidate="novalidate" data-msg="Please select the documents you have" action="/select-documents" accept-charset="UTF-8" method="post"><input name="utf8" type="hidden" value="&#x2713;" /><input type="hidden" name="authenticity_token" value="2Vb46CN8Bljm/f6lHNxqu3PGWMVAsVdjIe6uJrnzoxbUic0vi8jU/Ea8UrmOWBtwan860qN2uvdFNW8DoFNVuQ==" />' +
        '<div class="form-group">' +
        '<h2 class="heading-medium">Do you have a valid UK photocard driving licence, full or provisional?</h2>' +
        '<div class="form-group form-field">' +
        '<fieldset>' +
        '<label class="block-label selection-button-radio" for="select_documents_form_any_driving_licence_true"><input type="radio" value="true" name="select_documents_form[any_driving_licence]" id="select_documents_form_any_driving_licence_true" /> Yes</label>' +
        '<label class="block-label selection-button-radio" for="select_documents_form_any_driving_licence_false"><input type="radio" value="false" name="select_documents_form[any_driving_licence]" id="select_documents_form_any_driving_licence_false" /> No</label>' +
        '</fieldset>' +
        '</div>' +
        '<div id="driving_licence_details" class="form-group panel panel-border-narrow">' +
        '<fieldset>' +
        '<legend><span class="form-label">Where was your driving licence issued?</span></legend>' +
        '<div class="multiple-choice"><input value="great_britain" name="select_documents_form[driving_licence]" id="select_documents_form_driving_licence_great_britain" type="radio"> <label for="select_documents_form_driving_licence_great_britain">Great Britain</label></div>' +
        '<div class="multiple-choice"><input value="northern_ireland" name="select_documents_form[driving_licence]" id="select_documents_form_driving_licence_northern_ireland" type="radio"> <label for="select_documents_form_driving_licence_northern_ireland">Northern Ireland</label></div>' +
        '</fieldset>' +
        '</div>' +
        '<h2 class="heading-medium">Do you have a UK passport?</h2>' +
        '<div class="form-group form-field">' +
        '<fieldset>' +
        '<label class="block-label selection-button-radio" for="select_documents_form_passport_true"><input type="radio" value="true" name="select_documents_form[passport]" id="select_documents_form_passport_true" /> Yes</label>' +
        '<label class="block-label selection-button-radio" for="select_documents_form_passport_yes_expired"><input type="radio" value="yes_expired" name="select_documents_form[passport]" id="select_documents_form_passport_yes_expired" /> Yes, but it&#39;s expired</label>' +
        '<label class="block-label selection-button-radio" for="select_documents_form_passport_false"><input type="radio" value="false" name="select_documents_form[passport]" id="select_documents_form_passport_false" /> No</label>' +
        '</fieldset>' +
        '</div>' +
        '</div>' +
        '<div id="passport_details" class="form-group panel panel-border-narrow js-hidden" data-msg="Please enter valid passport expiry date">' +
        '<fieldset>' +
          '<legend>' +
            '<span class="form-label">Depending on when your passport expired, you may still be able to use it.</span>' +
            '<span class="form-label-bold">When did your passport expire?</span>' +
            '<span class="form-hint">For example, 31 3 2016</span>' +
          '</legend>' +
          '<div class="form-date expiry-details">' +
            '<div class="form-group">' +
              '<label class="form-label" for="select_documents_form_passport_expiry_day">Day</label>' +
              '<input class="form-control" type="number" pattern="[0-9]*" max="31" name="select_documents_form[passport_expiry][day]" id="select_documents_form_passport_expiry_day" />' +
            '</div>' +
            '<div class="form-group">' +
              '<label class="form-label" for="select_documents_form_passport_expiry_month">Month</label>' +
              '<input class="form-control" type="number" pattern="[0-9]*" min="1" max="12" name="select_documents_form[passport_expiry][month]" id="select_documents_form_passport_expiry_month" />' +
            '</div>' +
            '<div class="form-group form-group-year" >' +
              '<label class="form-label" for="select_documents_form_passport_expiry_year">Year</label>' +
              '<input class="form-control" type="number" pattern="[0-9]*" min="0" max="2017" name="select_documents_form[passport_expiry][year]" id="select_documents_form_passport_expiry_year" />' +
            '</div>' +
          '</div>' +
        '</fieldset>' +
        '</div>' +
        '<div id="validation-error-message-js"></div>' +
        '<div class="actions">' +
          '<input type="submit" name="commit" value="Continue" class="button" id="next-button" />' +
          '<a href="/other-identity-documents">I don&#39;t have either of these documents</a>' +
        '</div>' +
        '</form>';


    var selectDocumentsForm;
    var $dom;

    function answerQuestion(document, answer) {
        selectDocumentsForm.find('input[name="select_documents_form[' + document + ']"][value=' + answer + ']')
            .prop('checked', true)
            .trigger('click');
    }
    
    function submitForm() {
        selectDocumentsForm.triggerHandler('submit')
    }

    function expectErrorMessage(error) {
        expect(selectDocumentsForm.find('#validation-error-message-js').text()).toBe(error);
    }

    function expectErrorMessageToContain(error) {
        expect(selectDocumentsForm.find('#validation-error-message-js').text()).toContain(error);
    }

    function expectNoError() {
        expect(selectDocumentsForm.children('.form-group:first').is('.error')).toBe(false);
        expect(selectDocumentsForm.find('#validation-error-message-js').text()).toBe('');
    }

    beforeEach(function () {
        $dom = $('<div>' + formWithNoErrors + '</div>');
        $(document.body).append($dom);
        GOVUK.validation.init();
        GOVUK.selectDocumentsCleverQuestions.init();
        selectDocumentsForm = GOVUK.selectDocumentsCleverQuestions.$form;
        this.selectNoPassport = function () {
            answerQuestion('passport', false);
        };
        this.selectPassportExpired = function () {
          answerQuestion('passport', 'yes_expired');
        };
        this.selectYesPassport = function () {
            answerQuestion('passport', true);
        };
        this.selectYesValidDrivingLicence = function () {
            answerQuestion('any_driving_licence', true);
        };
        this.selectNoValidDrivingLicence = function () {
            answerQuestion('any_driving_licence', false);
        };
        this.selectGBDrivingLicence = function () {
            answerQuestion('driving_licence', 'great_britain');
        };
        this.enterValidPassportExpiryDate = function (day, month, year) {
          selectDocumentsForm.find('#select_documents_form_passport_expiry_day').val(day).trigger('blur');
          selectDocumentsForm.find('#select_documents_form_passport_expiry_month').val(month).trigger('blur');
          selectDocumentsForm.find('#select_documents_form_passport_expiry_year').val(year).trigger('blur');
        }
    });

    afterEach(function () {
        $dom.remove();
    });

    it("should not initially show driving licence details section.", function () {
        expect(selectDocumentsForm.find('#driving_licence_details').attr("class")).toContain("js-hidden");
    });

    it("should show driving licence details section when user indicates valid driving licence", function () {
        this.selectYesValidDrivingLicence();
        expect(selectDocumentsForm.find('#driving_licence_details').attr("class")).not.toContain("js-hidden");
    });

    it("should not show driving licence details section when user indicates no valid driving licence", function () {
        this.selectNoValidDrivingLicence();
        expect(selectDocumentsForm.find('#driving_licence_details').attr("class")).toContain("js-hidden");
    });

    it("should show passport expiry details section when user selects passport has expired", function () {
        this.selectPassportExpired();
        expect(selectDocumentsForm.find('#passport_details').attr("class")).not.toContain("js-hidden");
    });

    it("should have no errors on initialising the form.", function () {
        expect(selectDocumentsForm.find('.error').length).toBe(0);
    });

    it("should have errors on submit when no selections made.", function () {
        submitForm();
        expectErrorMessage('Please select the documents you have');
    });

    it("should have errors on submit when no driving licence selection made.", function () {
        this.selectNoPassport();
        submitForm();
        expectErrorMessage('Please select the documents you have');
    });

    it("should have errors on submit when no passport selection made.", function () {
        this.selectNoValidDrivingLicence();
        submitForm();
        expectErrorMessage('Please select the documents you have');
    });

    it("should have errors on submit when no passport expiry date given when passport expiration selection made.", function () {
        this.selectNoValidDrivingLicence();
        this.selectPassportExpired();
        submitForm();
        expectErrorMessageToContain('Please enter valid passport expiry date');
    });

    it("should clear errors when driving licence and passport details selected", function () {
        submitForm();
        expectErrorMessage('Please select the documents you have');
        this.selectYesValidDrivingLicence();
        this.selectYesPassport();
        expectNoError();
    });

    it("should clear expiry errors when passport expiry details entered", function () {
        this.selectNoValidDrivingLicence();
        this.selectPassportExpired();
        submitForm();
        expectErrorMessageToContain('Please enter valid passport expiry date');
        this.enterValidPassportExpiryDate(1, 1, 1999);
        expectNoError();
    });

    it("should clear expiry errors when passport is selected to Yes", function () {
        this.selectNoValidDrivingLicence();
        this.selectPassportExpired();
        submitForm();
        expectErrorMessageToContain('Please enter valid passport expiry date');
        this.selectYesPassport();
        expectNoError();
      });

    it("should have error when neither GB or NI details are not given for driving licence", function () {
        this.selectYesValidDrivingLicence();
        submitForm();
        expectErrorMessage('Please select the documents you have');
    });

    it("should not ask for driving licence details when no driving licence is selected", function () {
        this.selectNoValidDrivingLicence();
        submitForm();
        expectErrorMessage('Please select the documents you have');
    });

    it("should clear errors when driving licence details are given", function () {
        this.selectYesValidDrivingLicence();
        this.selectYesPassport();
        submitForm();
        expectErrorMessage('Please select the documents you have');
        this.selectGBDrivingLicence();
        expectNoError();
    });
});
